### ISSUE TITLE
Argument type is partially unknown when using dict()

### ISSUE THREAD

**Issue by tekumara** [microsoft/pyright#847]
Given:
```python
from typing import Any, Dict


def config(user: str) -> Dict[str, Any]:
    config: Dict[str, Any] = dict()
    config['Instances'] = dict(
        TerminationProtected=False,
        Name="awesome",
        Location="mars"
    )
    return config
```

pyright is returning:
```
  6:5 - error: Argument type is partially unknown
  Argument corresponds to parameter "v" in function "__setitem__"
  Argument type is "dict[Unknown, bool | Literal['mars', 'awesome']]" (reportUnknownArgumentType)
```

pyright 1.1.52
Python 3.6.10

**Comment by erictraut** [comment]
This is the intended behavior. I presume you have enabled "strict" mode (or have specifically enabled the `reportUnknownArgumentType` diagnostic rule).

If you allocate a dict using `dict()`, the type arguments are unknown. In effect, it becomes a `Dict[str, Unknown]`, which means that the type checker has no visibility into what value types are permitted. In strict mode, all unspecified types are flagged as errors. To fix, you need to specify the type via an annotation, as you did in the first line of the function above.

Alternatively, if you use a dictionary expression rather than the `dict()` constructor, Pyright will infer the type of the dict according to the type inference rules specified [here](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#dictionary-expressions).

```
    config["Instances"] = {
        "TerminationProtected": False,
        "Name": "awesome",
        "Location": "mars",
    }
```

**Comment by tekumara** [comment]
Thanks for the detailed explanation! I was a little surprised because it looks like pyright can infer the dictionary values, but not the keys, ie: in the above example it detects  `dict[Unknown, bool | Literal['mars', 'awesome']]`

The other solution appears to be using `dict()` but extracting the value so a type annotation can be specified:
```python
    instances: Dict[str, Any] = dict(
        TerminationProtected=False,
        Name="awesome",
        Location="mars"
    )
    config['Instances'] = instances
```

**Comment by erictraut** [comment]
Pyright uses inference only in cases where type information is not provided. In the case of the `dict()` constructor, type information is provided in the `builtins.pyi` type stub. It contains three overloads. Note that two of the three overloads would allow a type checker to determine both `_KT` (the key type) and `_VT` (the value type). The first overload (which is the one that applies in your use case) allows for `_VT` only because `_KT` does not appear anywhere.

```
    @overload
    def __init__(self, **kwargs: _VT) -> None: ...
    @overload
    def __init__(self, map: Mapping[_KT, _VT], **kwargs: _VT) -> None: ...
    @overload
    def __init__(self, iterable: Iterable[Tuple[_KT, _VT]], **kwargs: _VT) -> None: ...
```

The solution you've suggested (using a variable type annotation) works great.

**Comment by tekumara** [comment]
Super explanation, thanks so much @erictraut that makes sense now!
